### PR TITLE
Golang docs

### PR DESCRIPTION
**Pull request by jackkleeman**
The best way to review this is just to look at the preview

### CODE REVIEW

File: docs/develop/go/overview.mdx [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+- We pass the struct to `restate.Object` instead of `restate.Service`.
+- The first argument of each handler must be the `ObjectContext` parameter.
+Handlers with the `ObjectContext` parameter can write to the K/V state store.
+Only one handler can be active at a time, to ensure consistency.
+- If you want to have a handler that executes concurrently to the others and
+doesn't have write access to the K/V state, instead use the
+`ObjectSharedContext`. For example, you can use these handlers to read K/V
+state, or interact with the blocking handler.
+
+<Admonition type="tip" title="Prefer a bit more type-safety?">
```
Comment (slinkydeveloper) [inline docs/develop/go/overview.mdx:66]: Using the term `type safety` here doesn't seem right to me, our api is already type safe... I would use something like "do you like to start from the service definition" or something like that.

File: docs/develop/go/service-communication.mdx [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+
+# Service Communication
+
+A handler can call another handler and wait for the response (request-response), or it can send a message without waiting for the response.
+
+## Request-response calls
+
+Request-response calls are requests where the client waits for the response.
+
+<CH.Scrollycoding className={"single-item"}>
```
Comment (slinkydeveloper) [inline docs/develop/go/service-communication.mdx:16]: there is a problem with tab here, not sure what

File: docs/develop/go/service-communication.mdx [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+
+    To a Service:
+
+    ```go
+    CODE_LOAD::go/develop/servicecommunication.go#request_response_service
```
Comment (slinkydeveloper) [inline docs/develop/go/service-communication.mdx:21]: All the snippets on this page are too large, and i have an horizontal scroll bar for those. Perhaps can you newline on the .Request/.Send?

File: docs/develop/go/journaling-results.mdx [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+## Selectors
+Operations such as calls, awakeables, and `ctx.After` return futures which can
+be safely selected over using `ctx.Select`. Restate will log the order in which they complete, to make this deterministic on replay.
+
+```go
+CODE_LOAD::go/develop/journalingresults.go#race
+```
+
+<Admonition type="caution" title={"Avoid using Goroutines"}>
```
Comment (slinkydeveloper) [inline docs/develop/go/journaling-results.mdx:48]: It's even stronger imo :)

```suggestion
<Admonition type="caution" title={"Don't use Goroutines!"}>
```

File: docs/develop/go/journaling-results.mdx [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+
+    ### Generating random numbers
+
+    Methods exist on `ctx.Rand()` for generating `float64` and `uint64`, or
+    otherwise `ctx.Rand().Source()` can be provided to `math/rand/v2` as a
+    source for any random operation.
+
+    Once the SDK is upgraded to rely on Go 1.22, `math/rand/v2` will be embedded
+    into the struct returned by `ctx.Rand()` and will be usable directly. Use of
+    the older `math/rand` package is not recommended.
```
Comment (slinkydeveloper) [inline docs/develop/go/journaling-results.mdx:80]: I think this text is something for the godocs, and not really for the manual, i would remove it.

**Review by slinkydeveloper — COMMENTED** [review]
Left some minor comments, from my side the code looks good!

File: docs/get_started/tour.mdx [hunk 6]
```diff
@@ -214,6 +248,115 @@ CartObject       1
...
+ExpireTicket  value of content-type 'application/json'  value of content-type 'application/json'
+Checkout      none                                      none
+AddTicket     value of content-type 'application/json'  value of content-type 'application/json'
+
+- TicketObject
+Type: VirtualObject ⬅️ 🚶🚶🚶
+HANDLER     INPUT  OUTPUT
+MarkAsSold  none   none
+Unreserve   none   none
+Reserve     none   none
```
Comment (gvdongen) [inline docs/get_started/tour.mdx:326]: Shouldn't this return a boolean? So `value of content-type 'application/json'`?

File: docs/get_started/tour.mdx [hunk 7]
```diff
@@ -214,6 +248,115 @@ CartObject       1
...
+- CheckoutService
+Type: Service
+HANDLER  INPUT                                     OUTPUT
+Handle   value of content-type 'application/json'  value of content-type 'application/json'
+
+- CartObject
+Type: VirtualObject ⬅️ 🚶🚶🚶
+HANDLER       INPUT                                     OUTPUT
+ExpireTicket  value of content-type 'application/json'  value of content-type 'application/json'
+Checkout      none                                      none
```
Comment (gvdongen) [inline docs/get_started/tour.mdx:318]: Shouldn't this return a boolean? So `value of content-type 'application/json'`?

File: docs/develop/go/service-communication.mdx (hunk 2 above)
Comment (jackkleeman) [reply docs/develop/go/service-communication.mdx:16]: not sure what you mean

Comment (slinkydeveloper) [reply docs/develop/go/service-communication.mdx:16]: Look here https://golang.documentation-beg.pages.dev/develop/go/service-communication#request-response-calls

**Comment by slinkydeveloper** [comment]
I have one general comment that the indentation is too much... e.g. this:

image

Compared to this:

image

This is especially visible in snippets with fairly long-ish lines... Can you configure go fmt to use 2 spaces instead of 4?

File: code_snippets/go/develop/state.go [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+	_ = stateKeys
+
+	// <start_get>
+	myString, err := restate.Get[string](ctx, "my-string-key")
+	if err != nil {
+		return err
+	}
+	if myString == "" {
+		myString = "my-default"
+	}
```
Comment (slinkydeveloper) [inline code_snippets/go/develop/state.go:25]: To make this case abundantly clear, i would use a pointer type instead.

**Comment by gvdongen** [comment]
I think the checking of the go code snippets still needs to be added to the github workflows as well

**Comment by gvdongen** [comment]
And we should add Go icons to the overview page here: https://golang.documentation-beg.pages.dev/

File: code_snippets/go/develop/errorhandling.go [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+	"fmt"
+
+	restate "github.com/restatedev/sdk-go"
+)
+
+type ErrorHanding struct{}
+
+func (ErrorHanding) Greet(ctx restate.Context, name string) error {
+	// <start_here>
+	return restate.TerminalError(fmt.Errorf("Something went wrong."), 500)
```
[the remark this answers is not in the document]
Comment (jackkleeman) [reply code_snippets/go/develop/errorhandling.go:13]: Nah that fn doesn't take a code. And I want to show how to give a code

File: code_snippets/go/develop/state.go (hunk 8 above)
Comment (jackkleeman) [reply code_snippets/go/develop/state.go:25]: I thought about this. But then you have to assign the default to a var and then take a reference to it, which is ugly. This is why people use proto.String. I will take another look

**Comment by jackkleeman** [comment]
@slinkydeveloper go uses tabs, i will check how to get our code blocks to display them tighter

**Pull request merged**